larrecodnn/ImagePatternAlgs/NuSonic/Triton/TritonData.cc

```diff
@@ -0,0 +1,185 @@
+void TritonInputData::toServer(std::shared_ptr<TritonInput<DT>> ptr) {
+  const auto& data_in = *ptr;
+
+  //check batch size
+  if (data_in.size() != batchSize_) {
+    throw cet::exception("TritonDataError") << name_ << " input(): input vector has size " << data_in.size()
+                                            << " but specified batch size is " << batchSize_;
+  }
+
+  //shape must be specified for variable dims or if batch size changes
+  data_->SetShape(fullShape_);
+
+  if (byteSize_ != sizeof(DT))
+    throw cet::exception("TritonDataError") << name_ << " input(): inconsistent byte size " << sizeof(DT)
+                                            << " (should be " << byteSize_ << " for " << dname_ << ")";
+
+  int64_t nInput = sizeShape();
+  for (unsigned i0 = 0; i0 < batchSize_; ++i0) {
+    const DT* arr = data_in[i0].data();
+    triton_utils::throwIfError(data_->AppendRaw(reinterpret_cast<const uint8_t*>(arr), nInput * byteSize_),
+                               name_ + " input(): unable to set data for batch entry " + std::to_string(i0));
+  }
+
+  //keep input data in scope
+  holder_ = std::move(ptr);
+}
+
+template <>
+template <typename DT>
+TritonOutput<DT> TritonOutputData::fromServer() const {
+  if (!result_) {
+    throw cet::exception("TritonDataError") << name_ << " output(): missing result";
+  }
+
+  if (byteSize_ != sizeof(DT)) {
+    throw cet::exception("TritonDataError") << name_ << " output(): inconsistent byte size " << sizeof(DT)
+                                            << " (should be " << byteSize_ << " for " << dname_ << ")";
+  }
+
+  uint64_t nOutput = sizeShape();
+  TritonOutput<DT> dataOut;
+  const uint8_t* r0;
+  size_t contentByteSize;
+  size_t expectedContentByteSize = nOutput * byteSize_ * batchSize_;
+  triton_utils::throwIfError(result_->RawData(name_, &r0, &contentByteSize), "output(): unable to get raw");
+  if (contentByteSize != expectedContentByteSize) {
+    throw cet::exception("TritonDataError") << name_ << " output(): unexpected content byte size " << contentByteSize
+                                            << " (expected " << expectedContentByteSize << ")";
+  }
+
+  const DT* r1 = reinterpret_cast<const DT*>(r0);
+  dataOut.reserve(batchSize_);
+  for (unsigned i0 = 0; i0 < batchSize_; ++i0) {
+    auto offset = i0 * nOutput;
+    dataOut.emplace_back(r1 + offset, r1 + offset + nOutput);
+  }
+
+  return dataOut;
+}
+
+template <>
+void TritonInputData::reset() {
+  data_->Reset();
+  holder_.reset();
+}
+
+template <>
+void TritonOutputData::reset() {
+  result_.reset();
+}
+
+//explicit template instantiation declarations
+template class TritonData<nic::InferInput>;
+template class TritonData<nic::InferRequestedOutput>;
+
+template void TritonInputData::toServer(std::shared_ptr<TritonInput<float>> data_in);
+template void TritonInputData::toServer(std::shared_ptr<TritonInput<int64_t>> data_in);
+
+template TritonOutput<float> TritonOutputData::fromServer() const;
```

Comment:
You should be able to remove these explicit instantiations.

Reply:
get error when I remove these